crates/ethereum/engine/src/orchestrator.rs

```diff
@@ -0,0 +1,224 @@
+        from_tree: mpsc::UnboundedReceiver<EngineApiEvent>,
+    ) -> Self {
+        self.tree_channels = Some(TreeChannels { to_tree, from_tree });
+        self
+    }
+
+    fn with_pipeline_and_spawner(
+        mut self,
+        pipeline: Pipeline<DB>,
+        pipeline_task_spawner: Box<dyn TaskSpawner>,
+    ) -> Self {
+        self.pipeline_container = Some(PipelineContainer { pipeline, pipeline_task_spawner });
+        self
+    }
+
+    fn build(self) -> eyre::Result<EthChainOrchestrator<DB, Client>> {
+        let (
+            Some(chain_spec),
+            Some(client),
+            Some(incoming_requests),
+            Some(tree_channels),
+            Some(pipeline_container),
+        ) = (
+            self.chain_spec,
+            self.client,
+            self.incoming_requests,
+            self.tree_channels,
+            self.pipeline_container,
+        )
+        else {
+            eyre::bail!("not all required compoenents specified")
+        };
+
+        let consensus = Arc::new(EthBeaconConsensus::new(chain_spec));
+        let downloader = BasicBlockDownloader::new(client, consensus);
+
+        let TreeChannels { to_tree, from_tree } = tree_channels;
+        let engine_handler = EngineApiRequestHandler::new(to_tree, from_tree);
+        let handler = EngineHandler::new(engine_handler, downloader, incoming_requests);
+
+        let PipelineContainer { pipeline, pipeline_task_spawner } = pipeline_container;
+        let backfill_sync = PipelineSync::new(pipeline, pipeline_task_spawner);
+
+        Ok(EthChainOrchestrator { orchestrator: ChainOrchestrator::new(handler, backfill_sync) })
+    }
+}
+
+/// The type that drives the Ethereum chain forward and communicates progress.
+#[pin_project]
+#[allow(missing_debug_implementations)]
+pub struct EthChainOrchestrator<DB, Client>
+where
+    DB: Database + 'static,
+    Client: HeadersClient + BodiesClient + Clone + Unpin + 'static,
+{
+    orchestrator: EthChainOrchestratorType<DB, Client>,
+}
+
+impl<DB, Client> Future for EthChainOrchestrator<DB, Client>
+where
+    DB: Database + 'static,
+    Client: HeadersClient + BodiesClient + Clone + Unpin + 'static,
+{
+    type Output = Result<(), EthChainOrchestratorError>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        // Call poll on the inner orchestrator.
+        match self.project().orchestrator.poll_next_unpin(cx) {
+            Poll::Ready(Some(_event)) => Poll::Ready(Ok(())),
+            Poll::Ready(None) => Poll::Ready(Ok(())),
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+/// Potential error returned by `EthChainOrchestrator`.
+#[derive(Debug)]
+pub struct EthChainOrchestratorError {}
+
+/// Helper container for tree channels in `EthChainOrchestratorBuilder`.
+struct TreeChannels<T: EngineTypes> {
+    to_tree: mpsc::Sender<FromEngine<BeaconEngineMessage<T>>>,
+    from_tree: mpsc::UnboundedReceiver<EngineApiEvent>,
+}
```

Comment:
I assume this helper type isn't actually required and we could directly use `EngineApiRequestHandler`?